### ISSUE TITLE
PrivateDevices=yes silently disables RootImage=

### ISSUE THREAD

**Issue by lucaswerkmeister** [systemd/systemd#9737]
**systemd version the issue has been seen with**
v239

**Used distribution**
 Arch Linux

**Expected behaviour you didn't see**
 I was experimenting with portable services, and tried to run a service which uses `PrivateDevices=yes` as a portable service. After attaching the service, I expected to be able to run the service normally.

**Unexpected behaviour you saw**
 Starting the service failed with the error:

```
FOO.service: Failed to execute command: No such file or directory
FOO.service: Failed at step EXEC spawning /usr/local/bin/FOO: No such file or directory
```

With log level set to `debug`, the following message also appears (produced [here](https://github.com/systemd/systemd/blob/v239/src/core/execute.c#L2385)):

> FOO.service: Failed to set up namespace, assuming containerized execution, ignoring: Operation not permitted

And `strace` revealed that systemd failed to open `/dev/loop-control`:

> `openat(AT_FDCWD, "/dev/loop-control", O_RDWR|O_NOCTTY|O_NONBLOCK|O_CLOEXEC)` = -1 EPERM (Operation not permitted)

**Steps to reproduce the problem**

```sh
sudo systemd-run -p PrivateDevices=yes -p RootImage=/var/lib/portables/FOO.raw -qt /bin/cat /etc/machine-id
```

This will print the machine ID of the host system, not of the portable image. Without the `-p PrivateDevices=yes`, it will print the machine ID of the image. (Note: non-bootable `mkosi` images have an empty machine ID, so there might be no output – you can try some other commands to differentiate the environments, e. g. `/bin/ls /`.)

~~If you don’t have a real OS image handy, any regular file will do for the first step, since it isn’t used anyways. (Personally, I find it amusing to use `/proc/sys/kernel/ostype` for such purposes.)~~

@poettering noticed that this only affects cgroups v1 systems.

**Comment by lucaswerkmeister** [comment]
I’m not sure *why* this issue occurs – as far as I can see, `apply_private_devices` happens *much* later in `exec_child` than `apply_mount_namespace`. But perhaps the fix will still be a simple reordering…

If it isn’t possible to fix this issue properly (and instead it would have to be documented as a known restriction in `systemd.exec(5)`), I believe we could still work around it for the portable images case by specifying `PrivateDevices=no` in the default profile. This should still be reasonably safe, since the default profile also specifies `DynamicUser=yes`.

**Comment by lucaswerkmeister** [comment]
> If you don’t have a real OS image handy, any regular file will do for the first step, since it isn’t used anyways. (Personally, I find it amusing to use /proc/sys/kernel/ostype for such purposes.)

Disregard that – using a non-image file will still print the machine ID of the host system (demonstrating the more general problem that some errors in namespace setup are ignored), but it’s due to a different problem, not a failure to allocate a loop device.

**Comment by lucaswerkmeister** [comment]
You can also play with this bug in an interactive session:

```sh
sudo systemd-run -p PrivateDevices=yes -p RootImage=/var/lib/portables/FOO.raw -qt bash
```

In this environment, you’re still in the real root file system, and even the `PrivateDev=yes` hasn’t been applied. You can reproduce the error from `strace` with

```sh
exec 3<>/dev/loop-control
bash: /dev/loop-control: Operation not permitted
```

And yet, I am unable to figure out where that error is coming from. `ls -l /proc/self/ns/` shows the same symlink targets (same inode numbers). `cat /proc/mounts` shows the same mounts. There are two capabilities missing from the full set (`CAP_SYS_RAWIO` and `CAP_MKNOD`), but removing them in an external session with `setpriv` doesn’t result in the same error.

**Comment by lucaswerkmeister** [comment]
It looks like we can’t open any but a select set of devices under `/dev/`. The ones we can open in this environment are:

```sh
success=();
for file in /dev/*; do
  if [[ -L $file ]] || ! [[ -b $file || -c $file ]]; then
    continue;
  fi;
  if { exec 3<>$file; } 2>/dev/null; then
    success+=("$file");
    exec 3<&-;
  fi;
done;
printf '* %s\n' "${success[@]}"
```

* /dev/full
* /dev/null
* /dev/ptmx
* /dev/random
* /dev/tty
* /dev/urandom
* /dev/zero

This list is suspiciously identical to the list of device nodes in systemd’s `mount_private_dev`, and some of those files (`ptmx`) really don’t have many other matches in the systemd code bases that could serve as alternative explanations… but still, as far as I can tell, `setup_namespace` sets up the root image long before setting up the private `/dev`. (But I might be mistaken, there’s some indirection going on that I don’t fully see through.)

**Comment by yuwata** [comment]
Fix is waiting in #9744.
